File: backend/app/dtos/preprocessing.py
"""
Preprocessing DTOs - Request/Response models for preprocessing operations.
"""

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class NormalizationPreviewRequest(BaseModel):
    """Request for normalization preview."""

    method: Literal["none", "minmax", "zscore", "robust", "log"] = Field(
        default="minmax", description="Normalization method to apply"
    )
    features: Optional[List[str]] = Field(
        default=None, description="Features to preview. If None, all numeric features will be used"
    )
    sample_size: int = Field(
        default=10, ge=1, le=100, description="Number of sample values to return"
    )


class FeatureStats(BaseModel):
    """Statistics for a feature."""

    min: float
    max: float
    mean: float
    std: float


class FeaturePreview(BaseModel):
    """Preview data for a single feature."""

    data_type: str
    original: Dict[str, Any] = Field(description="Original values: {sample: [...], stats: {...}}")
    transformed: Dict[str, Any] = Field(
        description="Transformed values: {sample: [...], stats: {...}}"
    )


class NormalizationPreviewResponse(BaseModel):
    """Response for normalization preview."""

    method: str
    version_id: str
    features: Dict[str, FeaturePreview]
    total_rows: int
    message: Optional[str] = None

File: frontend/src/lib/api/preprocessing.ts
import { api } from "./client";

export type NormalizationMethod = "none" | "minmax" | "zscore" | "robust" | "log";

export interface NormalizationPreviewRequest {
    method: NormalizationMethod;
    features?: string[];
    sample_size?: number;
}

export interface FeatureStats {
    min: number;
    max: number;
    mean: number;
    std: number;
}

export interface FeaturePreview {
    data_type: string;
    original: {
        sample: number[];
        stats: FeatureStats;
    };
    transformed: {
        sample: number[];
        stats: FeatureStats;
    };
}

export interface NormalizationPreviewResponse {
    method: string;
    version_id: string;
    features: Record<string, FeaturePreview>;
    total_rows: number;
    message?: string;
}

export const preprocessingApi = {
    /**
     * Preview normalization transformation on selected features.
     */
    async previewNormalization(
        datasetId: string,
        versionId: string,
        request: NormalizationPreviewRequest
    ): Promise<NormalizationPreviewResponse> {
        const response = await api.post<NormalizationPreviewResponse>(
            `/datasets/${datasetId}/versions/${versionId}/preprocess/preview`,
            request
        );
        return response.data;
    },
};

export default preprocessingApi;
